network/packethandlers.c

```diff
@@ -2,4 +2,598 @@
+    {
+        //Allocate memory for the arbitrary data and copy it into the new buffer.
+        //Note, if allocation fails we just dump the data
+        arbitraryData = calloc(arbitrarySize, 1);
+        if(arbitraryData != NULL)
+        {
+            memcpy(arbitraryData, rawPacket->data + 4, arbitrarySize);
+        }
+    }
+    unpacked->arbitrary = arbitraryData;
+    //Dump it in the packet
+    packet->data = (PTLVData_Base_t*)unpacked;
+    //We own the packet memory, free it
+    llnet_packet_free(rawPacket);
+    return packet;
+}
+
+PacketTLV_t* unpackStateUpdate(IntermediateTLV_t* rawPacket)
+{
+    PacketTLV_t* packet = createBasePacket(rawPacket);
+    if (packet == NULL)
+    {
+        //Free the raw packet
+        llnet_packet_free(rawPacket);
+        return NULL;
+    }
+
+    errno = 0;
+    PTLVData_STATE_UPDATE_t* unpacked = calloc(1,sizeof(PTLVData_STATE_UPDATE_t));
+    //Allocation failed
+    if(unpacked == NULL && errno != 0)
+    {
+        //packet allocation succeeded, so free that memory
+        free(packet);
+        //Free the raw packet
+        llnet_packet_free(rawPacket);
+        return NULL;
+    }
+    //Byte one is the state
+    unpacked->new_state = (RobotState_t)rawPacket->data[0];
+    //The reserved isn't currently used, but we can use it in the future, so unpack it
+    unpacked->reserved = (uint32_t)(rawPacket->data[1]) << 16 |
+                         (uint32_t)(rawPacket->data[2]) << 8 | rawPacket->data[3];
+    //Total packet size - defined data size
+    size_t arbitrarySize = rawPacket->length - 4;
+    void* arbitraryData = NULL;
+    //There's arbitrary data here!
+    if(arbitrarySize > 0)
+    {
+        //Allocate memory for the arbitrary data and copy it into the new buffer.
+        //Note, if allocation fails we just dump the data
+        arbitraryData = calloc(arbitrarySize, 1);
+        if(arbitraryData != NULL)
+        {
+            memcpy(arbitraryData, rawPacket->data + 4, arbitrarySize);
+        }
+    }
+    unpacked->arbitrary = arbitraryData;
+    //Dump it in the packet
+    packet->data = (PTLVData_Base_t*)unpacked;
+    //We own the packet memory, free it
+    llnet_packet_free(rawPacket);
+    return packet;
+}
+
+PacketTLV_t* unpackConfigRequest(IntermediateTLV_t* rawPacket)
+{
+    PacketTLV_t* packet = createBasePacket(rawPacket);
+    if (packet == NULL)
+    {
+        //Free the raw packet
+        llnet_packet_free(rawPacket);
+        return NULL;
+    }
+
+    errno = 0;
+    PTLVData_CONFIG_REQUEST_t* unpacked = calloc(1,sizeof(PTLVData_CONFIG_REQUEST_t));
+    //Allocation failed
+    if(unpacked == NULL && errno != 0)
+    {
+        //packet allocation succeeded, so free that memory
+        free(packet);
+        //Free the raw packet
+        llnet_packet_free(rawPacket);
+        return NULL;
+    }
+    packet->type = rawPacket->type;
+    packet->timestamp = rawPacket->timestamp;
+    packet->length = rawPacket->length;
```

Comment:
isn't this unnecessary, done by `createBasePacket(...)`?

Reply:
Well would you look at that